open_spiel/games/phantom_go.cc

```diff
@@ -0,0 +1,793 @@
+#include "open_spiel/spiel_utils.h"
+
+namespace open_spiel {
+namespace phantom_go {
+namespace {
+
+// Facts about the game
+const GameType kGameType{
+    /*short_name=*/"phantom_go",
+    /*long_name=*/"Phantom Go",
+                   GameType::Dynamics::kSequential,
+                   GameType::ChanceMode::kDeterministic,
+                   GameType::Information::kImperfectInformation,
+                   GameType::Utility::kZeroSum,
+                   GameType::RewardModel::kTerminal,
+    /*max_num_players=*/2,
+    /*min_num_players=*/2,
+    /*provides_information_state_string=*/true,
+    /*provides_information_state_tensor=*/false,
+    /*provides_observation_string=*/true,
+    /*provides_observation_tensor=*/true,
+    /*parameter_specification=*/
+                   {{"komi", GameParameter(7.5)},
+                    {"board_size", GameParameter(9)},
+                    {"handicap", GameParameter(0)},
+                       // After the maximum game length, the game will end arbitrarily and the
+                       // score is computed as usual (i.e. number of stones + komi).
+                       // It's advised to only use shorter games to compute win-rates.
+                       // When not provided, it defaults to DefaultMaxGameLength(board_size)
+                    {"max_game_length",
+                     GameParameter(GameParameter::Type::kInt, /*is_mandatory=*/false)}},
+};
+
+std::shared_ptr<const Game> Factory(const GameParameters &params) {
+    return std::shared_ptr<const Game>(new PhantomGoGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+std::vector<VirtualPoint> HandicapStones(int num_handicap) {
+    if (num_handicap < 2 || num_handicap > 9) return {};
+
+    static std::array<VirtualPoint, 9> placement = {
+        {MakePoint("d4"), MakePoint("q16"), MakePoint("d16"), MakePoint("q4"),
+         MakePoint("d10"), MakePoint("q10"), MakePoint("k4"), MakePoint("k16"),
+         MakePoint("k10")}};
+    static VirtualPoint center = MakePoint("k10");
+
+    std::vector<VirtualPoint> points;
+    points.reserve(num_handicap);
+    for (int i = 0; i < num_handicap; ++i) {
+        points.push_back(placement[i]);
+    }
+
+    if (num_handicap >= 5 && num_handicap % 2 == 1) {
+        points[num_handicap - 1] = center;
+    }
+
+    return points;
+}
+
+}  // namespace
+
+PhantomGoState::PhantomGoState(std::shared_ptr<const Game> game, int board_size, float komi,
+                               int handicap)
+    : State(std::move(game)),
+      board_(board_size),
+      komi_(komi),
+      handicap_(handicap),
+      max_game_length_(game_->MaxGameLength()),
+      to_play_(GoColor::kBlack) {
+    ResetBoard();
+
+}
+
+//This method is used, when the Metapositon Resampling fails
+//It resamples the state into a Metaposition, that corresponds to the actual state on the game board
+std::unique_ptr<State> PhantomGoState::ResampleFromMetapositionHard(
```

Comment:
Do you want to keep these implementations? I thought you are doing it via the CSP and this is obsolete.